pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -82,11 +82,32 @@ public void add(MessageImpl<?> msg, SendCallback callback) {
     }
 
     private ByteBuf getCompressedBatchMetadataAndPayload() {
-        for (MessageImpl<?> msg : messages) {
+        batchedMessageMetadataAndPayload.markWriterIndex();
+        batchedMessageMetadataAndPayload.markReaderIndex();
```

Comment:
Instead of marking the index of the buffer, I think we can store the indexes into local variables.

The primary reason, is that when we pass the buffer to another function, we don't know whether  it will call `markWriterIndex()` again, and these calls are not stacked.

Using local variables should avoid this possible trap down the road.